internal/fxlog/spy.go

```diff
@@ -23,40 +23,63 @@ package fxlog
 import (
 	"fmt"
 	"strings"
+
+	"go.uber.org/zap"
 )
 
 // Spy is an Fx Printer that captures logged statements. It may be used in
 // tests of Fx logs.
 type Spy struct {
-	msgs []string
+	entries []Entry
 }
 
-var _ Printer = &Spy{}
+var _ Logger = &Spy{}
 
-// Printf logs the given message, formatting it in printf-style.
-func (s *Spy) Printf(msg string, args ...interface{}) {
-	s.msgs = append(s.msgs, fmt.Sprintf(msg, args...))
+// Log append an Entry.
+func (s *Spy) Log(entry Entry) {
+	s.entries = append(s.entries, entry)
 }
 
 // Messages returns a copy of all captured messages.
-func (s *Spy) Messages() []string {
-	msgs := make([]string, len(s.msgs))
-	copy(msgs, s.msgs)
+func (s *Spy) Messages() []Entry {
+	msgs := make([]Entry, len(s.entries))
+	copy(msgs, s.entries)
 	return msgs
 }
 
 // String returns all logged messages as a single newline-delimited string.
 func (s *Spy) String() string {
-	if len(s.msgs) == 0 {
+	if len(s.entries) == 0 {
 		return ""
 	}
+	var msg strings.Builder
+
+	for _, m := range s.entries {
+		_, _ = fmt.Fprintf(&msg, m.Message)
+		for _, f := range m.Fields {
+			// extra space before f.Key to separate out message.
+			_, _ = fmt.Fprintf(&msg, "\t%s: %v", f.Key, f.Value)
+		}
+		if m.Stack != "" {
+			_, _ = fmt.Fprintf(&msg, "%q", m.Stack)
+		}
+		_, _ = fmt.Fprintf(&msg, "\n")
```

Comment:
since we're ignoring the errors (writing to strings.builder cannot fail), we don't need to do the `_, _ =`.
```suggestion
		fmt.Fprintf(&msg, m.Message)
		for _, f := range m.Fields {
			// extra space before f.Key to separate out message.
			fmt.Fprintf(&msg, "\t%s: %v", f.Key, f.Value)
		}
		if m.Stack != "" {
			fmt.Fprintf(&msg, "\t%q", m.Stack)
		}
		msg.WriteString("\n")
```